muxers/mplex/src/io.rs

```diff
@@ -0,0 +1,771 @@
+    pub fn poll_close_stream(&mut self, cx: &mut Context<'_>, id: LocalStreamId)
+        -> Poll<io::Result<()>>
+    {
+        self.guard_ok()?;
+
+        match self.open_substreams.get(&id) {
+            None | Some(SubstreamState::SendClosed) => Poll::Ready(Ok(())),
+            Some(&state) => {
+                ready!(self.poll_send_frame(cx, || Frame::Close { stream_id: id }))?;
+                if state == SubstreamState::Open {
+                    debug!("Closed substream {} (half-close)", id);
+                    self.open_substreams.insert(id, SubstreamState::SendClosed);
+                } else if state == SubstreamState::RecvClosed {
+                    debug!("Closed substream {}", id);
+                    self.open_substreams.remove(&id);
+                    let below_limit = self.open_substreams.len() == self.config.max_substreams - 1;
+                    if below_limit {
+                        ArcWake::wake_by_ref(&self.notifier_open);
+                    }
+                }
+                Poll::Ready(Ok(()))
+            }
+        }
+    }
+
+    /// Sends a (lazily constructed) mplex frame on the underlying I/O stream.
+    ///
+    /// The frame is only constructed if the underlying sink is ready to
+    /// send another frame.
+    fn poll_send_frame<F>(&mut self, cx: &mut Context<'_>, frame: F)
+        -> Poll<io::Result<()>>
+    where
+        F: FnOnce() -> Frame<LocalStreamId>
+    {
+        let waker = NotifierWrite::register(&self.notifier_write, cx.waker());
+        match ready!(self.io.poll_ready_unpin(&mut Context::from_waker(&waker))) {
+            Ok(()) => {
+                let frame = frame();
+                trace!("Sending {:?}", frame);
+                match self.io.start_send_unpin(frame) {
+                    Ok(()) => Poll::Ready(Ok(())),
+                    Err(e) => Poll::Ready(self.on_error(e))
+                }
+            },
+            Err(e) => Poll::Ready(self.on_error(e))
+        }
+    }
+
+    /// Reads the next frame chosen by the given selection function.
+    ///
+    /// Before polling the underlying I/O stream, the receive frame
+    /// buffer is consulted.
+    ///
+    /// Returns `Pending` if no frame satisfying the selection function
+    /// is available, but the stream or substream is still open or
+    /// half-closed (i.e. still readable). All skipped frames are buffered and
+    /// tasks that previously registered interest in the streams for which
+    /// the skipped frames were received are notified.
+    ///
+    /// If `Some` stream ID is given, `Ok(None)` signals EOF for
+    /// that substream. If `None` is given for the stream ID,
+    /// `Ok(None)` is never returned. If the underlying connection
+    /// encounters EOF, a corresponding error is returned.
+    fn poll_read_frame<F, O>(
+        &mut self,
+        cx: &mut Context<'_>,
+        stream_id: Option<LocalStreamId>,
+        mut select: F,
```

Comment:
If I understand correctly `poll_read_frame` is only called by `poll_next_stream` and `poll_read_stream` with the former passing `None` as `stream_id` and the latter passing `Some`. Both seem to pass static closures via the `select` argument to `poll_read_frame`.

Accepting a function in `poll_read_frame / select` allows very diverse use-cases for the method, but as far as I can tell there are only the two mentioned above. Is this flexibility worth the complexity? Can `poll_read_frame` not imply the intention of its caller through `stream_id` only?

Reply:
> Can poll_read_frame not imply the intention of its caller through stream_id only?

Not quite, because the `select` function also maps the `Frame` to some other type, in the same manner as a function passed to `filter_map`. However, I made an attempt at restructuring the read control flow for more clarity, avoiding the function parameter, making `poll_read_frame` return a `Frame` and making the control flow of the individual read operations clearer, albeit at a bit of repeating a bit of of the same control flow in two places. It does seem preferable to me though. See https://github.com/libp2p/rust-libp2p/pull/1769/commits/7bd149888467429897b2b2ff6596e12cb803d025. What do you think?

Reply:
I missed the `map`ing functionality, indeed. Thanks for clarifying.

I am in favor of https://github.com/libp2p/rust-libp2p/commit/7bd149888467429897b2b2ff6596e12cb803d025, thanks for the change.